test/tstDetailsTreeNodeLabelling.cpp

```diff
@@ -0,0 +1,204 @@
+/****************************************************************************
+ * Copyright (c) 2017-2021 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
+#include "ArborX_EnableViewComparison.hpp"
+#include <ArborX_DetailsTreeNodeLabelling.hpp>
+#include <ArborX_LinearBVH.hpp>
+
+#include "BoostTest_CUDA_clang_workarounds.hpp"
+#include <boost/test/unit_test.hpp>
+
+#include <random>
+
+template <typename DeviceType, typename T>
+auto toView(std::vector<T> const &v, std::string const &lbl = "")
+{
+  Kokkos::View<T *, DeviceType> view(lbl, v.size());
+  Kokkos::deep_copy(view, Kokkos::View<T const *, Kokkos::HostSpace,
+                                       Kokkos::MemoryTraits<Kokkos::Unmanaged>>(
+                              v.data(), v.size()));
+  return view;
+}
+
+template <class MemorySpace>
+struct MockBVH
+{
+  Kokkos::View<Kokkos::pair<int, int> *, MemorySpace> children_;
+  KOKKOS_FUNCTION auto size() const { return children_.extent(0) + 1; }
```

Comment:
Just want to point out that one should be really careful about using this structure. The failing mode of `findParents` with wrong `children_` would either crash hard, or hang indefinitely.

The correct data for it should still form a tree.